### ISSUE TITLE
Similarity between documents?

### ISSUE THREAD

**Issue by inexxt** [RaRe-Technologies/gensim#836]
What exactly is the procedure in calculating similarity between documents?
As I understand from docs, the algorithm uses cosine similairty  

Running
```
from sklearn.metrics.pairwise import cosine_similarity

i1 = 10 # some index
i2 = 11 # some other index
l1 = docsdf["prep_content_base"][i1] # some sentence
l2 = docsdf["prep_content_base"][i2] # some other sentence

# I've used the same preprocessing during training
print(model.n_similarity(simple_preprocess(l1), 
                         simple_preprocess(l2))) 

v1 = model.infer_vector(simple_preprocess(l1))
v2 = model.infer_vector(simple_preprocess(l2))

print(cosine_similarity(v1, v2))

v1 = model.syn0norm[i1]
v2 = model.syn0norm[i2]

print(cosine_similarity(v1, v2))

v1 = model.docvecs[i1]
v2 = model.docvecs[i2]

print(cosine_similarity(v1, v2))
```
I get
```
0.171490121234
[[ 0.07252139]]
[[-0.02121822]]
[[-0.04322384]]
```
Which score is the correct one? (I need that number to cluster the documents set afterwards). What exactly are they representing?

**Comment by gojomo** [comment]
The 2nd and 4th numbers are each the similarity between two 'Paragraph Vector' doc-vecs for the two texts. The 2nd should be roughly equivalent to what you'd also get from:

    model.docvecs.similarity_unseen_docs(simple_preprocess(l1), simple_preprocess(l2))

Why are the 2nd and 4th numbers different? Both the initial training and later inference aren't deterministic, unless you try really hard to eliminate all sources of randomness, and even then, later-inference of a single vector from a frozen model won't be identical to the same text's results from prior bulk-training interleaved with all documents. Many find it useful to do many more than the default (5) number of inference 'steps'. See #374 & #515 for discussion of related issues. 

The 1st number, using `model.n_similarity()`, may be interesting for other reasons. It operates on `model` directly, whose comparison methods are all inherited from class Word2Vec. So, it actually uses only word-vectors. It averages together the vectors in each word-list, unit-normalizes them, then takes their cosine-similarity. This average-of-words document-vector is also often useful, but isn't what's calculated by PV-Doc2Vec. 

The 3rd number is unrelated nonsense. The `model.syn0norm` values are for word-vectors, not doc-vectors, so accessing them by doc-indexes i1/i2 gets random words. (And, some Doc2Vec modes, like plain DBOW – `dm=0` – don't train word-vecs at all, so any word-vec values in that case are still their random originally-initialized values.) To get at raw trained doc-vecs, look in `model.docvecs.doctag_syn0norm`. 

The better place for questions/discussion is the [project discussion list](https://groups.google.com/forum/#!forum/gensim) (rather than this bug/feature issue tracker).

**Comment by inexxt** [comment]
Thanks, and sorry for cluttering the tracker.

**Comment by gojomo** [comment]
No worries! Reach me on the list for any followups.
